[TOPIC-BLE_LLCP] Bluetooth: controller: Integrate phy update procedure

erbr-ot opened the pull request:
Adding payload data to phy update pdu
Integrating phy update procedure 
Misc. util functionality that should probably end up in ull_connections.c
Example of conditional compile of ull_llcp_phy.c in "min used channels" unittest
EDTT tests re. PHY update passes

erbr-ot committed:
bluetooth: controller: update data structures and conditional compile

Update phy update related data structures and setup and use conditional
compile of ull_llcp_phy.c

Signed-off-by: Erik Brockhoff <erbr@demant.com>

remove unused entry

erbr-ot committed:
bluetooth: controller: update api for phy update procedure

Signed-off-by: Erik Brockhoff <erbr@demant.com>

erbr-ot committed:
tests: bluetooth: controller: phy update pdu helpers update

fill in and check data in pdu helpers

Signed-off-by: Erik Brockhoff <erbr@demant.com>

erbr-ot committed:
bluetooth: controller: integrate phy update procedure

Integrating phy update with controller
Introducing tx-ack pdu flow
Updating pdu encode/decode

Signed-off-by: Erik BRockhoff <erbr@demant.com>

update comments

erbr-ot commented:
rebase after encryption PR merge

erbr-ot commented on tests/bluetooth/controller/common/src/helper_pdu.c:
```diff
@@ -180,23 +180,30 @@ void helper_pdu_encode_reject_ext_ind(struct pdu_data *pdu, void *param)
 	pdu->len = offsetof(struct pdu_data_llctrl, phy_req) + sizeof(struct pdu_data_llctrl_phy_req);
 	pdu->llctrl.opcode = PDU_DATA_LLCTRL_TYPE_PHY_REQ;
-	/* TODO(thoh): Fill in correct data */
+	if (!param) {
+		return;
+	}
```
Well, yes but it also allows to do unittests that do not consider payload, but only pdu flow, once the payload handling is implemented.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -84,17 +91,86 @@ enum {
- * LLCP Local Procedure PHY Update FSM
- */
+/* Hardcoded instant delta +6 */
+#define PHY_UPDATE_INSTANT_DELTA  6
+#define GENERATE_NTF  		1U
+#define DONT_GENERATE_NTF 	0U
```
yeap - space vs tab

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.c:
```diff
@@ -454,7 +454,7 @@ uint8_t ull_cp_encryption_pause(struct ll_conn *conn, const uint8_t rand[8], con
 }
 
-uint8_t ull_cp_phy_update(struct ll_conn *conn)
+uint8_t ull_cp_phy_update(struct ll_conn *conn, uint8_t tx, uint8_t flags, uint8_t rx, uint8_t host_cmd)
```
yeap

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -84,17 +91,86 @@ enum {
+
+static inline uint8_t select_phy(uint8_t _x)
+{
+	/* select only one phy, prefer 2M */
```
Suggestion added

erbr-ot replied:
re-written

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -153,18 +315,25 @@ static void lp_pu_ntf(struct ll_conn *conn, struct proc_ctx *ctx)
 	pdu = (struct node_rx_pu *)ntf->pdu;
 
 	pdu->status = ctx->data.pu.error;
+	pdu->rx = conn->lll.phy_rx;
+	pdu->tx = conn->lll.phy_tx;
 
 	/* Enqueue notification towards LL */
 	ll_rx_put(ntf->hdr.link, ntf);
 	ll_rx_sched();
 }
 
-static void lp_pu_complete(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
+static void lp_pu_complete(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, uint8_t ntf)
```
suggestion added

erbr-ot replied:
re-written

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn_types.h:
```diff
@@ -419,6 +419,15 @@ struct llcp_struct {
 	} cu;
+#if defined(CONFIG_BT_CTLR_PHY)
+	struct {
+		uint8_t tx:3;
+		uint8_t rx:3;
+		uint8_t flags:1;
+		uint8_t cmd:1;
+		uint8_t disabled:1;
```
only disabled flag is used the rest is removed

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:
```diff
@@ -132,8 +132,14 @@ struct proc_ctx {
+			uint8_t tx:3;
+			uint8_t rx:3;
+			uint8_t flags:1;
+			uint8_t cmd:1;
```
yeap

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:
```diff
@@ -954,25 +967,39 @@ static inline void pdu_encode_reject_ext_ind(struct pdu_data *pdu, uint8_t rejec
  * PHY Update Procedure Helper
  */
-void ull_cp_priv_pdu_encode_phy_req(struct pdu_data *pdu);
+void ull_cp_priv_pdu_encode_phy_req(struct pdu_data *pdu, uint8_t rx_phys, uint8_t tx_phys);
```
Hmm, yeah, well, I noticed the reject_ext_ind above, which did a similar 'exception to the rule' and since encode_req and encode_rsp need to use data from ctx and conn respectively, I figured it'd be cleaner to do it this way?

erbr-ot replied:
I updated as you suggested. It looks nicer, thnx

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_internal.h:
```diff
@@ -954,25 +967,39 @@ static inline void pdu_encode_reject_ext_ind(struct pdu_data *pdu, uint8_t rejec
+	return ull_cp_priv_pdu_encode_phy_req(pdu, rx_phys, tx_phys);
+}
+
+void ull_cp_priv_pdu_encode_phy_rsp(struct pdu_data *pdu, uint8_t rx_phys, uint8_t tx_phys);
```
see above

erbr-ot replied:
same as above

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -84,17 +91,86 @@ enum {
-static uint16_t lp_event_counter(struct ll_conn *conn)
+static void util_reset_timing_restrict(struct ll_conn *conn)
+{
+	struct lll_conn *lll = &conn->lll;
+
+	lll->phy_tx_time = conn->lll.phy_tx;
```
good point

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -84,17 +91,86 @@ enum {
+#define GENERATE_NTF  		1U
+#define DONT_GENERATE_NTF 	0U
+
+static inline uint8_t select_phy(uint8_t _x)
```
co__me_oN ;-)

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -84,17 +91,86 @@ enum {
+			 * will be restricted to fit current
+			 * connEffectiveMaxTxTime.
+			 */
+			uint8_t phy_tx_time[8] = {PHY_1M, PHY_1M, PHY_2M,
```
No, I guess not - I inherited this code snippet. I guess it could be optimized a tiny bit, however, it would only make it less readable in my opinion.

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -84,17 +91,86 @@ enum {
+	}
+}
+
+static void util_prep_update_ind(struct ll_conn *conn, struct proc_ctx *ctx)
```
agree

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -110,6 +186,88 @@ static uint16_t lp_event_counter(struct ll_conn *conn)
 	return event_counter;
 }
 
+static inline uint8_t util_instant_passed(struct ll_conn *conn, uint16_t instant)
+{
+	// if instant already passed, quit procedure with error */
+	uint16_t event_counter = util_event_counter(conn);
+	if (((event_counter - instant) & 0xFFFF) <= 0x7FFF) {
+		return 1;
+	}
+	return 0;
+}
```
thnx

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -110,6 +186,88 @@ static uint16_t lp_event_counter(struct ll_conn *conn)
+{
+	conn->phy_pref_rx = rx;
+	conn->phy_pref_tx = tx;
+	conn->lll.phy_flags = flags;
```
I inherited this too, and I had the same thought but decided after looking into it a while back, that it is indeed correct. However when I look at it now, I can't seem to remember why I thought it was ok! phy_pref_flags was introduced by me when I thought I needed it, so either I do need it or it should be removed again. 
Perhaps we can discuss tomorrow?

erbr-ot replied:
Ahh, I remembered again - sleep is good:
Since 'flags' indicate local coded phy preference (S2 or S8) and this is not negotiated with the peer, it is simply re-configured when the update is initiated, and takes effect whenever the coded phy is in use. So all is good and I remove the 'void' phy_pref_flags entry

thoh-ot replied:
Add a comment in the code about this

erbr-ot replied:
added comment

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -270,17 +460,50 @@ static void lp_pu_st_wait_tx_phy_update_ind(struct ll_conn *conn, struct proc_ct
+		const uint8_t end_procedure = util_check_update_ind(conn, ctx);
+		if (!end_procedure)
+		{
+			tx_resume_data(conn);
```
I believe you are correct - I'd tried to port the functionality from legacy llcp, in which case there appears to be a llcp_phy.pause_tx = 0 in the rx phy update ind case, however I went over the flow again in the legacy impl. and I believe there would already have been a llcp_phy.pause_tx = 0 in tx_ack of RSP

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -270,17 +460,50 @@ static void lp_pu_st_wait_tx_phy_update_ind(struct ll_conn *conn, struct proc_ct
+			lp_pu_complete(conn, ctx, evt, ctx->data.pu.cmd);
+		}
+
+		tx_resume_data(conn);
```
also 'cleaned' up after the above mentioned eval

erbr-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp.h:
```diff
@@ -94,7 +94,7 @@ void ull_cp_ltk_req_neq_reply(struct ll_conn *conn);
  * @brief Initiate a PHY Update Procedure.
  */
-uint8_t ull_cp_phy_update(struct ll_conn *conn);
+uint8_t ull_cp_phy_update(struct ll_conn *conn,uint8_t tx, uint8_t flags, uint8_t rx, uint8_t host_cmd);
```
yeap

erbr-ot committed:
bluetooth: controller: generate dle notification on timing changes

When a phy update results in a change in effective rx/tx timing
a DLE notification should be signaled host-ward

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

erbr-ot committed:
tests: bluetooth: controller: adding tests for tx data pause/resume

Adding checks to unittests to confirm proper tx data pause/resume
control in phy update procedure

Updateing commenting style

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -421,47 +729,47 @@ static void rp_pu_tx(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t opcode)
-	ll_rx_put(ntf->hdr.link, ntf);
-	ll_rx_sched();
-}
+	// when complete reset timing restrictions - idempotent (so no problem if we need to wait for NTF)
```
Use /**/

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -506,6 +813,10 @@ static void rp_pu_st_idle(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t ev
 static void rp_pu_st_wait_rx_phy_req(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
 {
+	pdu_decode_phy_req(ctx, (struct pdu_data *)param);
+	// Combine with the 'Prefered' the phys in conn->phy_pref_?x
```
Use /**/

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -538,6 +849,38 @@ static void rp_pu_st_wait_tx_phy_rsp(struct ll_conn *conn, struct proc_ctx *ctx,
+	case RP_PU_EVT_ACK:
+		if (ctx->state == RP_PU_STATE_WAIT_TX_ACK_PHY_RSP) {
+			LL_ASSERT(conn->lll.role == BT_HCI_ROLE_SLAVE);
+			// When we act as peripheral apply timing restriction
```
Use /**/

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -538,6 +849,38 @@ static void rp_pu_st_wait_tx_phy_rsp(struct ll_conn *conn, struct proc_ctx *ctx,
+			LL_ASSERT(conn->lll.role == BT_HCI_ROLE_SLAVE);
+			// When we act as peripheral apply timing restriction
+			pu_set_timing_restrict(conn, pu_select_phy_timing_restrict(conn, ctx->data.pu.tx));
+			// RSP acked, now await update ind from master
```
Use /**/

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -538,6 +849,38 @@ static void rp_pu_st_wait_tx_phy_rsp(struct ll_conn *conn, struct proc_ctx *ctx,
+		} else if (ctx->state == RP_PU_STATE_WAIT_TX_ACK_PHY_UPDATE_IND) {
+			LL_ASSERT(conn->lll.role == BT_HCI_ROLE_MASTER);
+			if (ctx->data.pu.m_to_s_phy || ctx->data.pu.s_to_m_phy) {
+				// UPDATE_IND acked, so lets await instant
```
Use /**/

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -538,6 +849,38 @@ static void rp_pu_st_wait_tx_phy_rsp(struct ll_conn *conn, struct proc_ctx *ctx,
+				// UPDATE_IND acked, so lets await instant
+				if (ctx->data.pu.m_to_s_phy)
+				{
+					// And if master to slave phys changes apply timining restrictions
```
Use /**/

thoh-ot commented on tests/bluetooth/controller/ctrl_phy_update/src/main.c:
```diff
@@ -716,6 +880,13 @@ void test_phy_update_sla_loc_collision(void)
+	struct pdu_data_llctrl_phy_upd_ind ind = { .instant = 7, .m_to_s_phy = PHY_2M, .s_to_m_phy = PHY_1M };
+	struct pdu_data_llctrl_length_rsp length_ntf = { 3*PDU_DC_PAYLOAD_SIZE_MIN, PKT_US(3*PDU_DC_PAYLOAD_SIZE_MIN,PHY_2M),
+							3*PDU_DC_PAYLOAD_SIZE_MIN, PKT_US(3*PDU_DC_PAYLOAD_SIZE_MIN,PHY_1M)};
+//	struct pdu_data_llctrl_phy_upd_ind ind_2 = { .instant = 14, .m_to_s_phy = PHY_2M, .s_to_m_phy = 0 };
```
Remove if not supposed to be used

thoh-ot commented on subsys/bluetooth/controller/ll_sw/ull_llcp_phy.c:
```diff
@@ -421,47 +731,46 @@ static void rp_pu_tx(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t opcode)
 static void rp_pu_complete(struct ll_conn *conn, struct proc_ctx *ctx, uint8_t evt, void *param)
 {
-	if (!ntf_alloc_is_available()) {
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+	const uint8_t ntf_count = ctx->data.pu.ntf_dle + ctx->data.pu.ntf_pu;
+#else
+	const uint8_t ntf_count = ctx->data.pu.ntf_pu;
+#endif
```
Make this one similar `lp_pu_complete`

thoh-ot reviewed (approved):
LGTM

erbr-ot commented:
rebased

kruithofa reviewed (approved):
LGTM
